components/zoho_creator/zoho_creator.app.mjs

```diff
@@ -1,11 +1,112 @@
+import { axios } from "@pipedreamhq/platform";
+
 export default {
   type: "app",
   app: "zoho_creator",
   propDefinitions: {},
   methods: {
-    // this.$auth contains connected account data
-    authKeys() {
-      console.log(Object.keys(this.$auth));
+    _authToken() {
+      return this.$auth.oauth_access_token;
+    },
+    _makeRequestConfig() {
+      const authToken = this._authToken();
+      const headers = {
+        "Authorization": `Zoho-oauthtoken ${authToken}`,
+        "User-Agent": "@PipedreamHQ/pipedream v0.1",
+      };
+      return {
+        headers,
+      };
+    },
+    _apiUrl() {
+      return `https://creator.${this.$auth.base_api_uri}/api/v2`;
+    },
+    _apiUsername() {
+      return this.$auth.oauth_uid;
+    },
+    _applicationsUrl() {
+      const baseUrl = this._apiUrl();
+      return `${baseUrl}/applications`;
+    },
+    _reportsUrl(appLinkName) {
+      const baseUrl = this._apiUrl();
+      const username = this._apiUsername();
+      return `${baseUrl}/${username}/${appLinkName}/reports`;
+    },
+    _reportDetailsUrl(appLinkName, reportLinkName) {
+      const baseUrl = this._apiUrl();
+      const username = this._apiUsername();
+      return `${baseUrl}/${username}/${appLinkName}/report/${reportLinkName}`;
+    },
+    recordsPageSize() {
+      return 200;
+    },
+    computeLastRecordsPage({
+      count = 0,
+      pageSize = this.recordsPageSize(),
+    }) {
+      return 1 + Math.floor(count / pageSize);
+    },
+    computeRecordsOffset({
+      count = 0,
+      pageSize = this.recordsPageSize(),
+    }) {
+      return count % pageSize;
+    },
+    async genericApiGetCall(url, params = {}) {
+      const baseRequestConfig = this._makeRequestConfig();
+      const requestConfig = {
+        ...baseRequestConfig,
+        params,
+        url,
+      };
+      return await axios(this, requestConfig);
+    },
+    async getApplications() {
+      const url = this._applicationsUrl();
+      const { applications } = await this.genericApiGetCall(url);
+      return applications;
+    },
+    async getReports() {
+      const applications = await this.getApplications();
+      const reports = [];
+      for (const app of applications) {
+        const url = this._reportsUrl(app.link_name);
+        const appReports = await this.genericApiGetCall(url);
+        appReports.reports.forEach((report) => {
+          report.app_link_name = app.link_name;
+          reports.push(report);
+        });
+      }
+      return reports;
+    },
+    async *getReportRecords( report, { page = 1 }) {
```

Comment:
Hi @dylburger, I would really appreciate fresh eyes looking into this function, this function paginates over the records in a report, but for some reason I can not trace an issue.
Context:
This generator should be fetching 200 records per page until it reaches the end where the API returns 404 error, hence try - catch block.

Reply:
I'm not sure if this is the issue, but it seems like you're using a [`page`](https://github.com/PipedreamHQ/pipedream/pull/1881/files/1d71bbaa37729857b24463933f6e972f2a50fdad#diff-33ec34c85e814c4085e5167f66d74341cb35f8949402c31b92bb81afce2bcd52R104) parameter in the request for pagination. And the [Get Records endpoint](https://www.zoho.com/creator/help/api/v2/get-records.html) uses a `from` parameter for pagination instead?

Reply:
Thanks, that must be it.